inference-engine/src/transformations/src/transformations/depth_to_space_fusion.cpp

```diff
@@ -0,0 +1,166 @@
+    uint64_t spatial_dims = shape_input.size() - 2;
+    possible_block_size = shape_reshape_before[2];
+    if (possible_block_size == 0)
+        return false;
+    uint64_t c_dim = shape_input[1] / std::pow(possible_block_size, spatial_dims);
+
+    // x' = reshape(data, [N, C / (block_size ^ K), block_size, block_size, ..., block_size, D1, D2, ..., DK])
+    ngraph::Shape expected_shape = {shape_input[0], c_dim};
+    for (uint64_t i = 0; i < spatial_dims; ++i)
+        expected_shape.push_back(possible_block_size);
+    for (uint64_t i = 2; i < shape_input.size(); ++i)
+        expected_shape.push_back(shape_input[i]);
+    is_transformation_valid &= (expected_shape == shape_reshape_before);
+
+    // x'' = transpose(x', [0,  1,  K + 2, 2, K + 3, 3, K + 4, 4, ..., K + (K + 1), K + 1])
+    ngraph::AxisVector expected_permutation = {0, 1};
+    for (uint64_t i = 2; i < shape_input.size(); ++i) {
+        expected_permutation.push_back(spatial_dims + i);
+        expected_permutation.push_back(i);
+    }
+    is_transformation_valid &= (expected_permutation == permutation);
+
+    // y = reshape(x'', [N, C / (block_size ^ K), D1 * block_size, D2 * block_size, D3 * block_size, ..., DK * block_size])
+    expected_shape = {shape_input[0], c_dim};
+    for (uint64_t i = 2; i < shape_input.size(); ++i)
+        expected_shape.push_back(shape_input[i] * possible_block_size);
+    is_transformation_valid &= (expected_shape == shape_reshape_after);
+
+    return is_transformation_valid;
+}
+
+void ngraph::pass::DepthToSpaceFusion::depth_to_space_fusion() {
+    auto input0 = std::make_shared<pattern::op::Label>(element::f32, Shape{1, 1, 1, 1});
+    auto input1 = std::make_shared<pattern::op::Label>(element::i64, Shape{4});
+    auto input2 = std::make_shared<pattern::op::Label>(element::i64, Shape{4});
+    auto input3 = std::make_shared<pattern::op::Label>(element::i64, Shape{4});
+    auto reshape_before = std::make_shared<ngraph::opset3::Reshape> (input0, input1, false);
+    auto permute = std::make_shared<ngraph::opset3::Transpose> (reshape_before, input2);
+    auto reshape_after = std::make_shared<ngraph::opset3::Reshape> (permute, input3, false);
+
+    ngraph::graph_rewrite_callback callback = [this](pattern::Matcher& m) {
+        // disable the transformation for all plugins, except CPU
+        if (!transformation_callback(std::make_shared<ngraph::opset3::DepthToSpace>())) {
+            return false;
+        }
+
+        auto reshape_after = std::dynamic_pointer_cast<ngraph::opset3::Reshape>(m.get_match_root());
+        if (!reshape_after) {
+            return false;
+        }
+
+        auto permute = std::dynamic_pointer_cast<ngraph::opset3::Transpose>(reshape_after->input_value(0).get_node_shared_ptr());
+        if (!permute) {
+            return false;
+        }
+
+        auto reshape_before = std::dynamic_pointer_cast<ngraph::opset3::Reshape>(permute->input_value(0).get_node_shared_ptr());
```

Comment:
You have pattern: ReshapeBegin->Transpose->ReshapeEnd here and you also need to check that ReshapeBegin and Transpose has exactly one consumer. Otherwise you can break a network.

Reply:
And please add negative test for this case.

Reply:
the test has been implemented